### PR TITLE
FPLA-2830: Record of NOC and Respondent details for a case after submission

### PR DESCRIPTION
**Pull request by samuelzerozero**
## JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-2830

### Change description ###
FPLA-2830: Record of NOC and Respondent details for a case after submission

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfChangeAboutToStartControllerTest.java [hunk 1]
```diff
@@ -37,26 +44,43 @@
...
         .builder()
         .build();
+    private static final String SECOND_RESPONDENT_FIRST_NAME = "Second";
+    private static final String SECOND_RESPONDENT_LAST_NAME = "Respondent";
+
+    @MockBean
+    private IdentityService identityService;
+
+    @MockBean
+    private Time time;
```
[the remark this answers is not in the document]
Comment (samuelzerozero) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfChangeAboutToStartControllerTest.java:83]: It feels like a bit more explicit to mock the time rather to assume that is the fixed current time

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/noc/ChangeOfRepresentationMethod.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package uk.gov.hmcts.reform.fpl.model.noc;
+
+
+public enum ChangeOfRepresentationMethod {
+    NOC("Noc"),
+    RESPONDENTS_EVENT("FPLA");
+
+    private String label;
+
+    ChangeOfRepresentationMethod(String label) {
```
[the remark this answers is not in the document]
Comment (samuelzerozero) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/noc/ChangeOfRepresentationMethod.java:10]: 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfChangeController.java [hunk 3]
```diff
@@ -34,7 +34,7 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody Cal
         CaseDetails caseDetails = callbackRequest.getCaseDetails();
         CaseData caseData = getCaseData(caseDetails);
 
-        caseDetails.getData().put("respondents1", noticeOfChangeService.updateRepresentation(caseData));
+        caseDetails.getData().putAll(noticeOfChangeService.updateRepresentation(caseData));
```
[the remark this answers is not in the document]
Comment (samuelzerozero) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfChangeController.java:37]: 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/RespondentAfterSubmissionRepresentationService.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                    .addedRepresentative(getSolicitor(respondentsAfter, changeOrganisationRequest))
+                    .removedRepresentative(getSolicitor(respondentsBefore, changeOrganisationRequest))
+                    .build());
+    }
+
+    private RespondentSolicitor getSolicitor(List<Element<Respondent>> respondents,
+                                             ChangeOrganisationRequest changeOrganisationRequest) {
+        int solicitorIdx = changeOrganisationRequest.getCaseRole().getIndex();
+
+        if (respondents == null || solicitorIdx >= respondents.size()) {
```
[the remark this answers is not in the document]
Comment (samuelzerozero) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/RespondentAfterSubmissionRepresentationService.java:75]: 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/RespondentAfterSubmissionRepresentationService.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+        final List<Element<Respondent>> respondentsAfter = defaultIfNull(caseData.getRespondents1(), emptyList());
+        final List<Element<Respondent>> respondentsBefore = defaultIfNull(caseDataBefore.getRespondents1(),
+            emptyList());
+
+        List<ChangeOrganisationRequest> representationChanges = respondentService.getRepresentationChanges(
+            respondentsAfter,
+            respondentsBefore);
+
+        updatedFields.put("changeOfRepresentatives", representationChanges.stream().sequential().reduce(
```
[the remark this answers is not in the document]
Comment (samuelzerozero) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/RespondentAfterSubmissionRepresentationService.java:47]: 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/representative/ChangeOfRepresentationRequest.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import uk.gov.hmcts.reform.fpl.model.RespondentSolicitor;
+import uk.gov.hmcts.reform.fpl.model.common.Element;
+import uk.gov.hmcts.reform.fpl.model.noc.ChangeOfRepresentation;
+import uk.gov.hmcts.reform.fpl.model.noc.ChangeOfRepresentationMethod;
+
+import java.util.List;
+
+@Value
+@Builder(toBuilder = true)
+public class ChangeOfRepresentationRequest {
```
[the remark this answers is not in the document]
Comment (samuelzerozero) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/representative/ChangeOfRepresentationRequest.java:15]: 👍

**Review by lewisbirks — APPROVED** [review]

File: ccd-definition/ComplexTypes/CareSupervision/NoticeOfChange/ChangeOfRepresenation.json [hunk 7]
```diff
@@ -0,0 +1,50 @@
+[
```
Comment (joez-f) [inline ccd-definition/ComplexTypes/CareSupervision/NoticeOfChange/ChangeOfRepresenation.json:1]: file name ChangeOfRepresenation -> ChangeOfRepresentation

Comment (samuelzerozero) [reply ccd-definition/ComplexTypes/CareSupervision/NoticeOfChange/ChangeOfRepresenation.json]: changed name 👍

**Review by tomaszpowroznik — APPROVED** [review]

**Pull request merged**